**2018.02.09 - prelim - disagreement analysis**

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Setup" data-toc-modified-id="Setup-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Setup</a></span><ul class="toc-item"><li><span><a href="#Setup---Imports" data-toc-modified-id="Setup---Imports-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Setup - Imports</a></span></li><li><span><a href="#Setup---virtualenv-jupyter-kernel" data-toc-modified-id="Setup---virtualenv-jupyter-kernel-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Setup - virtualenv jupyter kernel</a></span></li><li><span><a href="#Setup---Initialize-Django" data-toc-modified-id="Setup---Initialize-Django-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Setup - Initialize Django</a></span></li></ul></li><li><span><a href="#Evaluating-disagreements" data-toc-modified-id="Evaluating-disagreements-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Evaluating disagreements</a></span><ul class="toc-item"><li><span><a href="#Disagreement-tracking-process" data-toc-modified-id="Disagreement-tracking-process-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Disagreement tracking process</a></span></li><li><span><a href="#Data-in-Reliability_Names_Evaluation" data-toc-modified-id="Data-in-Reliability_Names_Evaluation-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Data in <code>Reliability_Names_Evaluation</code></a></span><ul class="toc-item"><li><span><a href="#Overall-disagreement-log" data-toc-modified-id="Overall-disagreement-log-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Overall disagreement log</a></span><ul class="toc-item"><li><span><a href="#flag-duplicates" data-toc-modified-id="flag-duplicates-2.2.1.1"><span class="toc-item-num">2.2.1.1&nbsp;&nbsp;</span>flag duplicates</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#CURRENT" data-toc-modified-id="CURRENT-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>CURRENT</a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#review-tags" data-toc-modified-id="review-tags-3.0.0.1"><span class="toc-item-num">3.0.0.1&nbsp;&nbsp;</span>review tags</a></span></li></ul></li><li><span><a href="#Ground-truth-coding-fixed" data-toc-modified-id="Ground-truth-coding-fixed-3.0.1"><span class="toc-item-num">3.0.1&nbsp;&nbsp;</span>Ground truth coding fixed</a></span><ul class="toc-item"><li><span><a href="#Mark-all-ground-truth-updates-as-TODO" data-toc-modified-id="Mark-all-ground-truth-updates-as-TODO-3.0.1.1"><span class="toc-item-num">3.0.1.1&nbsp;&nbsp;</span>Mark all ground truth updates as TODO</a></span></li><li><span><a href="#Mark-all-ground-truth-updates-as-human-error" data-toc-modified-id="Mark-all-ground-truth-updates-as-human-error-3.0.1.2"><span class="toc-item-num">3.0.1.2&nbsp;&nbsp;</span>Mark all ground truth updates as human error</a></span></li><li><span><a href="#Count-ground-truth-updates" data-toc-modified-id="Count-ground-truth-updates-3.0.1.3"><span class="toc-item-num">3.0.1.3&nbsp;&nbsp;</span>Count ground truth updates</a></span></li></ul></li><li><span><a href="#Reliability_Names-records-merged" data-toc-modified-id="Reliability_Names-records-merged-3.0.2"><span class="toc-item-num">3.0.2&nbsp;&nbsp;</span>Reliability_Names records merged</a></span></li><li><span><a href="#Deleted-Reliability_Names-records" data-toc-modified-id="Deleted-Reliability_Names-records-3.0.3"><span class="toc-item-num">3.0.3&nbsp;&nbsp;</span>Deleted Reliability_Names records</a></span></li></ul></li><li><span><a href="#Evaluating-Disagreements---Human-Error" data-toc-modified-id="Evaluating-Disagreements---Human-Error-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Evaluating Disagreements - Human Error</a></span></li><li><span><a href="#Disagreements---Computer-Error" data-toc-modified-id="Disagreements---Computer-Error-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Disagreements - Computer Error</a></span></li></ul></li></ul></div>

# Setup

- Back to [Table of Contents](#Table-of-Contents)

## Setup - Imports

- Back to [Table of Contents](#Table-of-Contents)

In [12]:
import datetime
import six

print( "packages imported at " + str( datetime.datetime.now() ) )

packages imported at 2018-02-12 02:02:33.671218


## Setup - virtualenv jupyter kernel

- Back to [Table of Contents](#Table-of-Contents)

If you are using a virtualenv, make sure that you:

- have installed your virtualenv as a kernel.
- choose the kernel for your virtualenv as the kernel for your notebook (Kernel --> Change kernel).

Since I use a virtualenv, need to get that activated somehow inside this notebook.  One option is to run `../dev/wsgi.py` in this notebook, to configure the python environment manually as if you had activated the `sourcenet` virtualenv.  To do this, you'd make a code cell that contains:

    %run ../dev/wsgi.py
    
This is sketchy, however, because of the changes it makes to your Python environment within the context of whatever your current kernel is.  I'd worry about collisions with the actual Python 3 kernel.  Better, one can install their virtualenv as a separate kernel.  Steps:

- activate your virtualenv:

        workon sourcenet

- in your virtualenv, install the package `ipykernel`.

        pip install ipykernel

- use the ipykernel python program to install the current environment as a kernel:

        python -m ipykernel install --user --name <env_name> --display-name "<display_name>"
        
    `sourcenet` example:
    
        python -m ipykernel install --user --name sourcenet --display-name "sourcenet (Python 3)"
        
More details: [http://ipython.readthedocs.io/en/stable/install/kernel_install.html](http://ipython.readthedocs.io/en/stable/install/kernel_install.html)

In [13]:
%pwd

'/home/jonathanmorgan/work/sourcenet/django/research/work/msu_phd_work/methods'

## Setup - Initialize Django

- Back to [Table of Contents](#Table-of-Contents)

First, initialize my dev django project, so I can run code in this notebook that references my django models and can talk to the database using my project's settings.

In [14]:
%run django_init.py

django initialized at 2018-02-12 02:02:42.404986


In [15]:
# django imports
from sourcenet_analysis.models import Reliability_Names_Evaluation

# Evaluating disagreements

- Back to [Table of Contents](#Table-of-Contents)

Look at stats for disagreements and evaluation, including human and computer errors.

- Notebook with work details: [2017.07.01-work_log-prelim_month-evaluate_disagreements.ipynb](https://research.local:8000/user/jonathanmorgan/notebooks/work/sourcenet/django/research/work/msu_phd_work/methods/2017.07.01-work_log-prelim_month-evaluate_disagreements.ipynb)

Process: Look at each instance where there is a disagreement and make sure the human coding is correct.

Most are probably instances where the computer screwed up, but since we are calling this human coding "ground truth", want to winnow out as much human error as possible.

For each disagreement, to check for coder error (like just capturing a name part for a person whose full name was in the story), click the "Article ID" in the column that has a link to article ID. It will take you to a view of the article where all the people who coded the article are included, with each detection of a mention or quotation displayed next to the paragraph where the person was originally first detected.

If not human error, remove TODO tag, filling in details on the diagreement in the record in `Reliability_Names_Evaluation` for the removal of the tag (details: [Disagreement tracking process](#Disagreement-tracking-process)).

If human error:

- 1) look at all the disagreements for the article.
- 2) remove all TODO tags from all disagreements, and fill in details for each.
- 3) follow steps below to create ground_truth copy and fix it.
- 4) rebuild Reliability_Names for article and cleanup.
- 5) then, do any deletes or merges you need to do, so you only do them once.

Pull together some numbers and analysis from disagreement work:

- counts of disagreements.
- which were human and computer error.
- ratio of human error to machine error.
- proportion of human and machine errors.
- overall number of disagreements compared to all decisions.
- characterisation of potential for systematic issues (not as bad as I feared).


## Disagreement tracking process

- Back to [Table of Contents](#Table-of-Contents)

From [2017.07.01-work_log-prelim_month-evaluate_disagreements.ipynb - Disagreement resolution](https://research.local:8000/user/jonathanmorgan/notebooks/work/sourcenet/django/research/work/msu_phd_work/methods/2017.07.01-work_log-prelim_month-evaluate_disagreements.ipynb#Disagreement-resolution):

For each disagreement, click on the article ID link in the row to go to the article and check to see if the human coding for the disagreement in question is correct ( [http://research.local/sourcenet/sourcenet/article/article_data/view_with_text/](http://research.local/sourcenet/sourcenet/article/article_data/view_with_text/) ).

Once you've evaluated and verified the human coding, remove the "`TODO`" tag from the current record (either from the single-article view above if you've removed all disagreements, or from the disagreement view if not):

- Click the checkbox in the "**select**" column next to the record whose evaluation is complete.
- In the "**Reliability names action:**" field, select "_Remove tag(s) from selected_".
- In the "**Tag(s) - (comma-delimited):**" field, enter "_`TODO`_" (without the quotes).
- Click the "**Do Action**" button.
- This will also place information on the `Reliability_Names` record into a `Reliability_Names_Evaluation` record in the database.  The message that results from this action completing will include a link to the record (the first number in the output).  Click the link to open the record and update it with additional details.  Specifically:

    - status - status of human coder's coding:
        
        - If the human coder got it right, status is "CORRECT", even if OpenCalais had an egregious error.
        - If this is because the coding screen couldn't capture compound names initially, set status to "INCOMPLETE", set the status message to "SKIPPED because screen couldn't deal with compound names", put the compound name string in notes, and then add tag "compound_names".
        - if the OC coder had an issue because we had to smoosh all paragraphs together because it didn't deal well with HTML markup in the body of text it processed, set status to "CORRECT", set status message to "OC ERROR because of formatting", and then explain the problem in the notes.  If the article is a list or column with odd formatting, consider flagging the article for removal from the study.
        - if you have to update ground truth, set "`status`" to "ERROR".
        - else, use your best judgment.
            
    - if problems caused by automated coder error, click the "`is_automated_error`" checkbox.
    - update the "`status_message`" so it contains a brief description of what exactly happened (should have been mentioned, should have been quoted, missed the person entirely, etc.).
    - update "`Notes`" with more details.
    
        - If should have been quoted or mentioned, note the graf # and paragraph text of the paragraph that indicates this.
    
    - add "`Tags`" if appropriate (for sports articles, for example, add "sports" tag).

_**NOTE: Always remove TODO tag first, so you have a record of status for each Reliability Names record.  Then, once you've done that, you can merge, delete, etc.**_

## Data in `Reliability_Names_Evaluation`

- Back to [Table of Contents](#Table-of-Contents)

### Overall disagreement log

- Track each Reliability_Names record with disagreement that we evaluate (All "remove tags" events with label "prelim_month"):
- Moved to `Reliability_Names_Evaluation` table in django: [http://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?label=prelim_month&o=-1.7.8.3.5](http://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?label=prelim_month&o=-1.7.8.3.5)
- [All "remove tags" events with label "prelim_month"](https://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?event_type__exact=remove_tags&label=prelim_month&o=-1.7.8.3.5)
- 428 total records.
- of those, 13 are same person and article, but different `Reliability_Names` record, so disagreements that had to be corrected twice because of rebuilding `Reliability_Names` for the article (either human error, or something weird).  SQL:

        SELECT sarne1.person_name,
            sarne1.id,
            sarne1.status,
            sarne1.original_reliability_names_id,
            sarne1.is_duplicate,
            sarne2.is_duplicate,
            sarne2.id,
            sarne2.status,
            sarne2.original_reliability_names_id
        FROM sourcenet_analysis_reliability_names_evaluation AS sarne1,
            sourcenet_analysis_reliability_names_evaluation AS sarne2
        WHERE sarne1.id != sarne2.id
            AND sarne1.label = 'prelim_month'
            AND sarne2.label = 'prelim_month'
            AND sarne1.event_type = 'remove_tags'
            AND sarne2.event_type = 'remove_tags'
            AND sarne1.article_id = sarne2.article_id
            AND sarne1.person_name = sarne2.person_name
            AND sarne1.original_reliability_names_id != sarne2.original_reliability_names_id
            AND sarne2.original_reliability_names_id > sarne1.original_reliability_names_id
        ORDER BY sarne1.id ASC;

- So, 428 - 13 = 415 unique disagreements.
- Could regenerate Reliability_Names without `ground_truth` to look at original counts?  Should be able to...  Just need to make sure I remember all steps...

    - would need to clear single names, then I'd be left with disagreements.  Not worth it.

#### flag duplicates

- Back to [Table of Contents](#Table-of-Contents)

TODO:

- duplicates

    - Make sure that for each pair of duplicates, one has "is_duplicate" checked.
    - to start, mark all that have duplicates as "is_to_do" = True.


**Step 1: get IDs of records with duplicates**

    SELECT DISTINCT ( sarne1.id )
    FROM sourcenet_analysis_reliability_names_evaluation AS sarne1,
        sourcenet_analysis_reliability_names_evaluation AS sarne2
    WHERE sarne1.id != sarne2.id
        AND sarne1.label = 'prelim_month'
        AND sarne2.label = 'prelim_month'
        AND sarne1.event_type = 'remove_tags'
        AND sarne2.event_type = 'remove_tags'
        AND sarne1.article_id = sarne2.article_id
        AND sarne1.person_name = sarne2.person_name
        AND sarne1.original_reliability_names_id != sarne2.original_reliability_names_id
        AND sarne2.original_reliability_names_id > sarne1.original_reliability_names_id
    ORDER BY sarne1.id ASC;

and

    SELECT DISTINCT ( sarne2.id )
    FROM sourcenet_analysis_reliability_names_evaluation AS sarne1,
        sourcenet_analysis_reliability_names_evaluation AS sarne2
    WHERE sarne1.id != sarne2.id
        AND sarne1.label = 'prelim_month'
        AND sarne2.label = 'prelim_month'
        AND sarne1.event_type = 'remove_tags'
        AND sarne2.event_type = 'remove_tags'
        AND sarne1.article_id = sarne2.article_id
        AND sarne1.person_name = sarne2.person_name
        AND sarne1.original_reliability_names_id != sarne2.original_reliability_names_id
        AND sarne2.original_reliability_names_id > sarne1.original_reliability_names_id
    ORDER BY sarne2.id ASC;


In [ ]:
# Got IDs that contain duplicates, now tag them as todo
ids_to_process = []
ids_to_process.append( 15 )
ids_to_process.append( 33 )
ids_to_process.append( 75 )
ids_to_process.append( 405 )
ids_to_process.append( 435 )
ids_to_process.append( 512 )
ids_to_process.append( 556 )
ids_to_process.append( 586 )
ids_to_process.append( 610 )
ids_to_process.append( 620 )
ids_to_process.append( 635 )
ids_to_process.append( 646 )
ids_to_process.append( 16 )
ids_to_process.append( 34 )
ids_to_process.append( 76 )
ids_to_process.append( 407 )
ids_to_process.append( 432 )
ids_to_process.append( 513 )
ids_to_process.append( 517 )
ids_to_process.append( 558 )
ids_to_process.append( 596 )
ids_to_process.append( 611 )
ids_to_process.append( 619 )
ids_to_process.append( 637 )
ids_to_process.append( 651 )

# retrieve model instances.

# get all evaluation records with label = "prelim_month" and IDs in our list.
evaluation_qs = Reliability_Names_Evaluation.objects.filter( label = "prelim_month" )
evaluation_qs = evaluation_qs.filter( pk__in = ids_to_process )

# count?
eval_count = evaluation_qs.count()
print( "record count: {}".format( str( eval_count ) ) )

# loop, setting "is_to_do" to True on each and saving.
for current_eval in evaluation_qs:
    
    # set is_to_do to True and set work_status to "metadata_review".
    current_eval.is_to_do = True
    current_eval.work_status = "duplicate_processing"
    current_eval.save()
    
#-- END loop over QuerySet. --#

Turns out, many of these were a missed person by human coder, that, once fixed, revealed a problem with the automated coding.  So, many were actually not duplicates, they were two separate issues with the same person.

# CURRENT

#### review tags

- Back to [Table of Contents](#Table-of-Contents)

TODO:

- get all tags.
- make boolean columns for each tag.
- for each record, set flags based on tags.
- then, flag all as todo and "review_disagreements", and go through them all to set metadata flags, and pick good ones for writing about disagreements.

In [19]:
'''
django-taggit documentation: https://github.com/alex/django-taggit

Adding tags to a model:

    from django.db import models
    
    from taggit.managers import TaggableManager
    
    class Food(models.Model):
        # ... fields here
    
        tags = TaggableManager()

Interacting with a model that has tags:

    >>> apple = Food.objects.create(name="apple")
    >>> apple.tags.add("red", "green", "delicious")
    >>> apple.tags.all()
    [<Tag: red>, <Tag: green>, <Tag: delicious>]
    >>> apple.tags.remove("green")
    >>> apple.tags.all()
    [<Tag: red>, <Tag: delicious>]
    >>> Food.objects.filter(tags__name__in=["red"])
    [<Food: apple>, <Food: cherry>]
    
    # include only those with certain tags.
    #tags_in_list = [ "prelim_unit_test_001", "prelim_unit_test_002", "prelim_unit_test_003", "prelim_unit_test_004", "prelim_unit_test_005", "prelim_unit_test_006", "prelim_unit_test_007" ]
    tags_in_list = [ "grp_month", ]
    if ( len( tags_in_list ) > 0 ):
    
        # filter
        print( "filtering to just articles with tags: " + str( tags_in_list ) )
        grp_article_qs = grp_article_qs.filter( tags__name__in = tags_in_list )
        
    #-- END check to see if we have a specific list of tags we want to include --#

'''

# imports
from sourcenet_analysis.models import Reliability_Names_Evaluation

# declare variables
evaluation_qs = None
record_count = -1
record_counter = -1
current_record = None
tag_to_count_map = {}
tag_qs = None
tag_list = None
current_tag = ""
cleaned_tag = ""
current_count = -1
tag_count = -1
no_tags_list = []

# get all evaluation records with label = "prelim_month" and event_type = "remove_tags".
evaluation_qs = Reliability_Names_Evaluation.objects.filter( label = "prelim_month" )
evaluation_qs = evaluation_qs.filter( event_type = "remove_tags" )

# first, just make sure that worked.
record_count = evaluation_qs.count()

# Check count of articles retrieved.
print( "Got " + str( record_count ) + " evaluations records." )

# loop over evaluations.
no_tags_count = 0
for current_record in evaluation_qs:

    # get tags.
    # current_article.tags.add( tag_value )
    tag_qs = current_record.tags.all()
    
    # output the tags.
    #print( "- Tags for record " + str( current_record.id ) + " : " + str( tag_qs ) )
    
    # count tags
    tag_count = tag_qs.count()
    
    # got tags?
    if ( tag_count > 0 ):
    
        # loop over tags.
        for current_tag in tag_qs:

            # standardize
            cleaned_tag = str( current_tag )

            # to lower case
            cleaned_tag = cleaned_tag.lower()

            # strip()
            cleaned_tag = cleaned_tag.strip()

            # in map?  Get current count.
            current_count = 0
            if ( cleaned_tag in tag_to_count_map ): 

                # It is in map - get counter for it.
                current_count = tag_to_count_map.get( cleaned_tag, None )

            #-- END check to see if tag in map --#

            # increment count and store.
            current_count += 1
            tag_to_count_map[ cleaned_tag ] = current_count

        #-- END loop over tags --#
        
    else:
        
        # increment no_tag_counter bby 1.
        no_tags_list.append( current_record )
        
    #-- END check to see if tags or not --#
    
#-- END loop over records --#

# output number of tagless evaluations
no_tags_count = len( no_tags_list )
print( "--> Count of articles with no tags: {}".format( str( no_tags_count ) ) )

# output tags
key_view = six.viewkeys( tag_to_count_map )
tag_list = list( key_view )
tag_list.sort()
for tag_string in tag_list:

    # print each tag and its count.
    current_count = tag_to_count_map.get( tag_string, -1 )
    print( "- {} ( {} )".format( str( tag_string ), str( current_count ) ) )
    
#-- END print tags. --#

Got 428 evaluations records.
--> Count of articles with no tags: 188
- ambiguity ( 2 )
- ambiguous ( 10 )
- complex ( 23 )
- complex_title ( 1 )
- complex_titles ( 25 )
- compound_attribution ( 1 )
- compound_names ( 9 )
- contributed_to ( 2 )
- dictionary_error ( 20 )
- disambiguation ( 4 )
- editing_error ( 3 )
- error ( 204 )
- follow_on_attribution ( 24 )
- foreign_names ( 1 )
- gender_confusion ( 2 )
- initials ( 5 )
- interesting ( 198 )
- layout_or_design ( 3 )
- list ( 14 )
- lists ( 1 )
- lookup ( 5 )
- no_html ( 5 )
- non_news ( 12 )
- possessive ( 1 )
- pronoun_attribution ( 2 )
- pronouns ( 4 )
- proper_noun ( 2 )
- proper_nouns ( 38 )
- quote_distance ( 12 )
- said_verb ( 29 )
- second_hand ( 10 )
- short_n-gram ( 5 )
- software_error ( 1 )
- spanish ( 1 )
- sports ( 6 )
- straightforward ( 74 )
- title_prefix ( 8 )


Output:

--> Count of articles with no tags: 188

Got 428 evaluations records.

- ambiguity ( 2 )
- ambiguous ( 10 )
- complex ( 23 )
- complex_title ( 1 )
- complex_titles ( 25 )
- compound_attribution ( 1 )
- compound_names ( 9 )
- contributed_to ( 2 )
- dictionary_error ( 20 )
- disambiguation ( 4 )
- editing_error ( 3 )
- error ( 204 )
- follow_on_attribution ( 24 )
- foreign_names ( 1 )
- gender_confusion ( 2 )
- initials ( 5 )
- interesting ( 198 )
- layout_or_design ( 3 )
- list ( 14 )
- lists ( 1 )
- lookup ( 5 )
- no_html ( 5 )
- non_news ( 12 )
- possessive ( 1 )
- pronoun_attribution ( 2 )
- pronouns ( 4 )
- proper_noun ( 2 )
- proper_nouns ( 38 )
- quote_distance ( 12 )
- said_verb ( 29 )
- second_hand ( 10 )
- short_n-gram ( 5 )
- software_error ( 1 )
- spanish ( 1 )
- sports ( 6 )
- straightforward ( 74 )
- title_prefix ( 8 )

tags to create:

- is_ambiguous = ambiguous, ambiguity
- --> is_attribution_compound = compound_attribution
- --> is_attribution_follow_on = follow_on_attribution
- --> is_attribution_pronoun = pronoun_attribution
- --> is_attribution_second_hand = second_hand
- is_complex = complex
- --> is_compound_names = compound_names
- --> is_contributed_to (and is_subject_shb_author) = contributed_to
- --> is_dictionary_error = dictionary_error
- --> is_disambiguation = disambiguation
- --> is_editing_error = editing_error
- is_error = error
- --> is_foreign_names = foreign_names
- --> is_gender_confusion = gender_confusion
- --> is_initials_error = initials
- is_interesting = interesting
- --> is_layout_or_design = layout_or_design
- is_list = list, lists
- --> is_lookup_error = lookup
- is_no_html = no_html
- is_not_hard_news = non_news
- --> is_possessive = possessive
- --> is_pronouns = pronouns
- --> is_proper_noun (and is_not_a_person) = proper_noun, proper_nouns
- --> is_quote_distance = quote_distance
- --> is_said_verb = said_verb
- --> is_short_n_gram = is_short_n-gram
- --> is_software_error = software_error
- --> is_spanish = spanish
- is_sports = sports
- is_straightforward = straightforward
- is_title = complex_title, complex_titles, title_prefix
- is_title_complex = complex_title, complex_titles
- is_title_prefix = title_prefix

Remember:

- If error ("-->"), make sure to set "is_error", as well.
- And, "is_automated_error" if automated error.

Notes:

- verified that all fields are in model.
- admin:

    - make sure that all fields are in admin.
    - reorganizing to make it a little less crazy.
    - make sure all are in limit list.
    
- Need to automatically set the flags based on the tag values.

In [11]:
'''
django-taggit documentation: https://github.com/alex/django-taggit

Adding tags to a model:

    from django.db import models
    
    from taggit.managers import TaggableManager
    
    class Food(models.Model):
        # ... fields here
    
        tags = TaggableManager()

Interacting with a model that has tags:

    >>> apple = Food.objects.create(name="apple")
    >>> apple.tags.add("red", "green", "delicious")
    >>> apple.tags.all()
    [<Tag: red>, <Tag: green>, <Tag: delicious>]
    >>> apple.tags.remove("green")
    >>> apple.tags.all()
    [<Tag: red>, <Tag: delicious>]
    >>> Food.objects.filter(tags__name__in=["red"])
    [<Food: apple>, <Food: cherry>]
    
    # include only those with certain tags.
    #tags_in_list = [ "prelim_unit_test_001", "prelim_unit_test_002", "prelim_unit_test_003", "prelim_unit_test_004", "prelim_unit_test_005", "prelim_unit_test_006", "prelim_unit_test_007" ]
    tags_in_list = [ "grp_month", ]
    if ( len( tags_in_list ) > 0 ):
    
        # filter
        print( "filtering to just articles with tags: " + str( tags_in_list ) )
        grp_article_qs = grp_article_qs.filter( tags__name__in = tags_in_list )
        
    #-- END check to see if we have a specific list of tags we want to include --#

'''

# imports
from sourcenet_analysis.models import Reliability_Names_Evaluation

# declare variables
evaluation_qs = None
record_count = -1
record_counter = -1
current_record = None
tag_set = set()
tag_qs = None
tag_list = None
current_tag = ""
cleaned_tag = ""

# get all evaluation records with label = "prelim_month" and event_type = "remove_tags".
evaluation_qs = Reliability_Names_Evaluation.objects.filter( label = "prelim_month" )
evaluation_qs = evaluation_qs.filter( event_type = "remove_tags" )

# first, just make sure that worked.
record_count = evaluation_qs.count()

# Check count of articles retrieved.
print( "Got " + str( record_count ) + " evaluations records." )

# loop over evaluations.
for current_record in evaluation_qs:

    # get tags.
    # current_article.tags.add( tag_value )
    tag_qs = current_record.tags.all()
    
    # output the tags.
    #print( "- Tags for record " + str( current_record.id ) + " : " + str( tag_qs ) )
    
    # loop over tags.
    for current_tag in tag_qs:
        
        # standardize
        cleaned_tag = str( current_tag )

        # to lower case
        cleaned_tag = cleaned_tag.lower()
        
        # strip()
        cleaned_tag = cleaned_tag.strip()
        
        # in set?
        if ( cleaned_tag not in tag_set ): 
        
            # If no, add to set.
            tag_set.add( cleaned_tag )
        
    #-- END loop over tags --#
    
#-- END loop over records --#

# output
tag_list = list( tag_set )
tag_list.sort()
for tag_string in tag_list:

    # print each tag.
    print( "- {}".format( str( tag_string ) ) )
    
#-- END print tags. --#

Got 428 evaluations records.
- ambiguity
- ambiguous
- complex
- complex_title
- complex_titles
- compound_attribution
- compound_names
- contributed_to
- dictionary_error
- disambiguation
- editing_error
- error
- follow_on_attribution
- foreign_names
- gender_confusion
- initials
- interesting
- layout_or_design
- list
- lists
- lookup
- no_html
- non_news
- possessive
- pronoun_attribution
- pronouns
- proper_noun
- proper_nouns
- quote_distance
- said_verb
- second_hand
- short_n-gram
- software_error
- spanish
- sports
- straightforward
- title_prefix


### Ground truth coding fixed

- For some, the error will be on the part of the human coder.  For human error, we create a new "`ground_truth`" record that we will correct, so we preserve original coding (and evidence of errors) in case we want or need that information later.  Below, we have a table of the articles where we had to fix ground truth.  To find the original coding, click the Article link.
- Denoted by records with "`is_ground_truth_fixed`" set to True in the `Reliability_Names_Evaluation` table in django:  [http://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?is_ground_truth_fixed__exact=1&label=prelim_month&o=-1.7.8.3.5](http://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?is_ground_truth_fixed__exact=1&label=prelim_month&o=-1.7.8.3.5)
- 130 total (130/415 = 31.3% - this is a lot - is this right?)

    - 4 duplicates, so call it 126.
    - based on "`prelim_month_human`" `Reliability_Names` tag, 135 disagreements between original and corrected coding.  Probably some merging needed here?
    - of those, 4 are same person and article, but different `Reliability_Names` record, so disagreements that had to be corrected twice because of rebuilding `Reliability_Names` for the article (either human error, or something else weird).  SQL:

            SELECT sarne1.person_name,
                sarne1.id,
                sarne1.status,
                sarne1.original_reliability_names_id,
                sarne1.article_id,
                sarne1.is_duplicate,
                sarne2.is_duplicate,
                sarne2.id,
                sarne2.status,
                sarne2.original_reliability_names_id,
                sarne2.article_id
            FROM sourcenet_analysis_reliability_names_evaluation AS sarne1,
                sourcenet_analysis_reliability_names_evaluation AS sarne2
            WHERE sarne1.id != sarne2.id
                AND sarne1.label = 'prelim_month'
                AND sarne2.label = 'prelim_month'
                AND sarne1.event_type = 'remove_tags'
                AND sarne2.event_type = 'remove_tags'
                AND sarne1.is_ground_truth_fixed = TRUE
                AND sarne2.is_ground_truth_fixed = TRUE
                AND sarne1.article_id = sarne2.article_id
                AND sarne1.person_name = sarne2.person_name
                AND sarne1.original_reliability_names_id != sarne2.original_reliability_names_id
                AND sarne2.original_reliability_names_id > sarne1.original_reliability_names_id
            ORDER BY sarne1.id ASC;

        Results (looks like these are ones that had to be merged, so ... minimze - when ambiguity, assume error in creating data, treat as duplicates so reduce count by number of duplicates):

| person_name | id | status | original_reliability_names_id | article_id | id | status | original_reliability_names_id | article_id |
|--|--|--|--|--|--|--|--|--|
| Jeff Hawkins | 33 | ERROR | 9408 | [21007](https://research.local/sourcenet/sourcenet/article/article_data/view_with_text/?article_id=21007) | 34 | ERROR | 9414 | 21007 |
| Fritz Wahlfield | 405 | ERROR | 10330 | [22415](https://research.local/sourcenet/sourcenet/article/article_data/view_with_text/?article_id=22415) | 407 | CORRECT | 10997 | 22415 |
| John Agar | 610 | ERROR | 8917 | [23904](https://research.local/sourcenet/sourcenet/article/article_data/view_with_text/?article_id=23904) | 611 | ERROR | 8918 | 23904 |
| Rachael Recker | 620 | ERROR | 8968 | [23920](https://research.local/sourcenet/sourcenet/article/article_data/view_with_text/?article_id=23920) | 619 | ERROR | 8976 | 23920 |

- work:

    - set all with "is_ground_truth_fixed" = True so "is_to_do" = True and "`work_status = "metadata_review"`".
    - update all with `is_ground_truth_updated = True` so `is_human_error = True`.


#### Mark all ground truth updates as TODO

- Back to [Table of Contents](#Table-of-Contents)

In [7]:
# get all evaluation records with label = "prelim_month", is_ground_truth_fixed = True, and event_type = "remove_tags".
evaluation_qs = Reliability_Names_Evaluation.objects.filter( label = "prelim_month" )
evaluation_qs = evaluation_qs.filter( is_ground_truth_fixed = True )
evaluation_qs = evaluation_qs.filter( event_type = "remove_tags" )

# count?
eval_count = evaluation_qs.count()
print( "record count: {}".format( str( eval_count ) ) )

# loop, setting "is_to_do" to True on each and saving.
for current_eval in evaluation_qs:
    
    # set is_to_do to True and set work_status to "metadata_review".
    current_eval.is_to_do = True
    current_eval.work_status = "metadata_review"
    current_eval.save()
    
#-- END loop over QuerySet. --#

record count: 130


#### Mark all ground truth updates as human error

- Back to [Table of Contents](#Table-of-Contents)

In [6]:
# get all evaluation records with label = "prelim_month", is_ground_truth_fixed = True, and event_type = "remove_tags".
evaluation_qs = Reliability_Names_Evaluation.objects.filter( label = "prelim_month" )
evaluation_qs = evaluation_qs.filter( is_ground_truth_fixed = True )
evaluation_qs = evaluation_qs.filter( event_type = "remove_tags" )

# count?
eval_count = evaluation_qs.count()
print( "record count: {}".format( str( eval_count ) ) )

# loop, setting "is_human_error" to True on each and saving.
for current_eval in evaluation_qs:
    
    # set is_to_do to True and set work_status to "metadata_review".
    current_eval.is_human_error = True
    current_eval.save()
    
#-- END loop over QuerySet. --#

record count: 127


#### Count ground truth updates

- Back to [Table of Contents](#Table-of-Contents)

Task 1: go through all "is_to_do" and update the metadata booleans.
    
- Link: [https://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?is_to_do__exact=1](https://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?is_to_do__exact=1)
- for duplicates, set "`is_duplicate`" to True on one of the two records.  If two records for a given person because of the computer finding a person and human missing them, mark the computer coder's evaluation record as the duplicate.
- for skipped, not a human error.


In [8]:
# get all evaluation records with:
# - label = "prelim_month"
# - is_ground_truth_fixed = True
# - event_type = "remove_tags"
# - is_duplicate = False
evaluation_qs = Reliability_Names_Evaluation.objects.filter( label = "prelim_month" )
evaluation_qs = evaluation_qs.filter( is_ground_truth_fixed = True )
evaluation_qs = evaluation_qs.filter( event_type = "remove_tags" )
# evaluation_qs = evaluation_qs.filter( is_to_do = True )  # 130 originally
evaluation_qs = evaluation_qs.filter( is_duplicate = False )  # now 123!

# count?
eval_count = evaluation_qs.count()
print( "record count: {}".format( str( eval_count ) ) )

# 123

record count: 123


Task 2: update counts of characterizations above.  When filtering for counts:
    
- filter out duplicates when generating counts (include only where "`is_duplicate`" = False).
- filter out skipped when generating counts (include only where "`is_skipped`" = False).
    


In [9]:
# get all evaluation records with:
# - label = "prelim_month"
# - is_ground_truth_fixed = True
# - is_human_error = True
# - event_type = "remove_tags"
# - is_duplicate = False
evaluation_qs = Reliability_Names_Evaluation.objects.filter( label = "prelim_month" )
evaluation_qs = evaluation_qs.filter( is_ground_truth_fixed = True )
evaluation_qs = evaluation_qs.filter( is_human_error = True )
evaluation_qs = evaluation_qs.filter( event_type = "remove_tags" )
# evaluation_qs = evaluation_qs.filter( is_to_do = True )  # 130 originally
evaluation_qs = evaluation_qs.filter( is_duplicate = False )  # now 123!

# count?
eval_count = evaluation_qs.count()
print( "record count: {}".format( str( eval_count ) ) )

# 102

record count: 102


Analysis:

- 130 total, 123 without duplicates, 102 if one removes those where human skipped a person because of limitation of the coding application (compound names).
- 102/415 = 31.3% - this is a lot - is this right?)
- number of affected articles?
- characterization of the problems:

    - is_missed_author:
    - is_missed_subject:
    - is_skipped (limitation of coding application):
    - is_author_shb_subject: *
    - is_subject_shb_author: *
    - is_quoted_shb_mentioned: *
    - is_mentioned_shb_quoted: *
    - is_wrong_text_captured: *
    - is_duplicate:
    
- note:

    - evaluation 469 is both author and subject.



TODO:

- work:

    - figure out number of affected articles (should just be count of Article_Data by ground_truth user).
    - update metadata for all disagreements (all "`remove_tags`" events).

### Reliability_Names records merged

- For some, need to merge a single-name detection by Calais with full-name detection by ground_truth (an OpenCalais error - did not detect full name - combined with lookup error - didn't lookup the right person since missed part of his or her name).  Will still have subsequently deleted one or more duplicate rows.
- Denoted by records with "`event_type`" set to "merge" in the Reliability_Names_Evaluation table in django: [http://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?event_type__exact=merge&label=prelim_month&o=-1.7.8.3.5](http://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?event_type__exact=merge&label=prelim_month&o=-1.7.8.3.5)
- 18 total
- TODO: need to check for the repeat thing SQL like in remove tags.


### Deleted Reliability_Names records

- Some records need to be deleted:

    - are just broken, need to be deleted.
    - cleanup after a merge (one stays, one goes).
    - cleanup after rebuilding `Reliability_Names` (single names removed once again, etc.).
    - ?

- Denoted by records with "`event_type`" set to "deleted" in the `Reliability_Names_Evaluation` table in django: [http://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?event_type__exact=delete&label=prelim_month&o=-1.7.8.3.5](http://research.local/sourcenet/admin/sourcenet_analysis/reliability_names_evaluation/?event_type__exact=delete&label=prelim_month&o=-1.7.8.3.5)
- 181 total
- TODO: need to provide examples.
- TODO: need to check for the repeat thing SQL like in remove tags.

In summary:

- after correction:

    - 2,446 overall coding decisions on people in "`prelim_month`".

        - https://research.local/sourcenet/sourcenet/analysis/reliability/names/disagreement/view
        - label: "`prelim_month`"
        - coders to compare (1 ==>): 2
        - Filter Type: "Lookup" (default)

    - 294 disagreements in "`prelim_month`".
    
        - https://research.local/sourcenet/sourcenet/analysis/reliability/names/disagreement/view
        - label: "`prelim_month`"
        - coders to compare (1 ==>): 2
        - Filter Type: "Disagree"

    - 135 disagreements between humans and corrected.
    
        - https://research.local/sourcenet/sourcenet/analysis/reliability/names/disagreement/view
        - label: "`prelim_month_human`"
        - coders to compare (1 ==>): 2
        - Filter Type: "Disagree"
        - to more readily explore distribution of problems, added tag "`prelim_month_human_disagree`"to disagreements in "`prelim_month_human`":
        
            - all disagreements:
            - label: "`prelim_month_human`"
            - coders to compare (1 ==>): 2
            - Filter Type: "Lookup"
            - Reliability_Names tags: "`prelim_month_human_disagree`"
            
- TODO:

    - look over human errors (disagreement between humans and corrected data).
    
        - Figure out types of error, counts of each type.  In SQL, filter on tag, then on other traits.
        
            - human miss: if coder 2 empty, 1 not, then human missed a person.
            - human false detect: if coder 1 empty, 2 not, human erroneously detected person.
            - 
            - 

## Evaluating Disagreements - Human Error

- Back to [Table of Contents](#Table-of-Contents)

Human error - Per article (how many have ground truth?) and per decision?  How many errors, compared to total number of decisions, and what kind of errors?

## Disagreements - Computer Error

- Back to [Table of Contents](#Table-of-Contents)

Computer error - look over classes of error for trends (systemic error) and interesting.